warp_prism/_warp_prism.c

```diff
@@ -12,6 +12,77 @@ const size_t signature_len = 11;
 const size_t column_buffer_growth_factor = 2;
 const size_t starting_column_buffer_length = 4096;
 
+#if __GNUC__ >= 5
+#define add_overflow __builtin_add_overflow
+#define mul_overflow __builtin_mul_overflow
+#else
+#define _overflow_opname(name, type) _ ## type ## _ ## name ## _overflow
+#define define_overflow_op(name, op, type)                              \
+    static inline bool _overflow_opname(name, type)(type a,             \
+                                                    type b,             \
+                                                    type* out) {        \
+        *out = a op b;                                                  \
+        /* This dumb jump-to-scalar instead of using `seto` allows gcc
+           to better optimize the function in the context that it is
+           used. We always do something like:
+
+           if (mul_overflow(a, b, out)) {
+               error_case()
+           }
+
+           By using jo here, gcc can fold the labels and turn this
+           into a jump right to `error_case()`. */                      \
+        __asm__ goto ("jo %l0\n" : : : : overflow);                     \
+        return false;                                                   \
+    overflow:                                                           \
+        return true;                                                    \
+    }
+
+
+define_overflow_op(add, +, int8_t)
+define_overflow_op(add, +, int16_t)
+define_overflow_op(add, +, int32_t)
+define_overflow_op(add, +, int64_t)
+define_overflow_op(add, +, uint8_t)
+define_overflow_op(add, +, uint16_t)
+define_overflow_op(add, +, uint32_t)
+define_overflow_op(add, +, uint64_t)
+
+define_overflow_op(mul, *, int8_t)
+define_overflow_op(mul, *, int16_t)
+define_overflow_op(mul, *, int32_t)
+define_overflow_op(mul, *, int64_t)
+define_overflow_op(mul, *, uint8_t)
+define_overflow_op(mul, *, uint16_t)
+define_overflow_op(mul, *, uint32_t)
+define_overflow_op(mul, *, uint64_t)
+
+#undef define_overflow_op
+
+#define _overflow_op_branch(name, type, alternative)                    \
+    __builtin_choose_expr(__builtin_types_compatible_p(__typeof__(*_out), type), \
+                          _overflow_opname(name, type)(_a, _b, (void*) _out), \
+                          alternative)
+
+#define _overflow_op(name, a, b, out) ({                         \
```

Comment:
I believe that this compiles.  That's about the extent of what I'm confident in asserting about what this does...